src/porepy/compositional/materials.py

```diff
@@ -0,0 +1,476 @@
+
+        The original constants in SI units are stored in :attr:`constants_in_SI`.
+
+        """
+
+        # Get the material constants using the data class field, excluding utilities
+        constants = asdict(self)
+        constants.pop("name")
+        constants.pop("units")
+        constants.pop("constants_in_SI")
+
+        # safety check if the user forgot to annotate the SI_units as a class variabel
+        # If yes, dataclass would declare it as a data field
+        if "SI_units" in constants:
+            raise AttributeError(
+                f"The attribute 'SI_units' in class {type(self)} must be annotated as"
+                + " SI_units: typing.ClassVar[dict[str,str]]."
+            )
+
+        # by logic, what remains are constants as numbers
+        # storing original constants
+        self.constants_in_SI.update(
+            cast(_HashableDict[str, number], _HashableDict(constants))
+        )
+
+        # transform the constants to SI and store them in this instance
+        # Bypass the freezing of attributes https://stackoverflow.com/questions/
+        # 53756788/how-to-set-the-value-of-dataclass-field-in-post-init-when-frozen-true
+        for k, v in self.constants_in_SI.items():
+            # convert the value to SI if it has a physical unit
+            if k not in type(self).SI_units:
+                raise AttributeError(
+                    f"Material constant {k} requires a declaration of its physical unit"
+                    + f" in SI in {type(self)}.SI_units."
+                )
+            si_unit = self.SI_units[k]
+            v_in_custom_units = self.units.convert_units(v, si_unit)
+            object.__setattr__(self, k, v_in_custom_units)
+
+    def __init_subclass__(cls) -> None:
+        """If a data-class inherits this class, the base class post-initialization
+        procedure must be inherited as well for the conversion from SI units
+        (at instantiation) to simulation :attr:`units`."""
+
+        if is_dataclass(cls):
+            if hasattr(cls, "__post_init__"):
+                if cls.__post_init__ != MaterialConstants.__post_init__:
+                    raise AttributeError(
+                        f"Data classes inheriting form {MaterialConstants} must not"
+                        + " have __post_init__ defined, but inherit it."
+                    )
+            else:
+                object.__setattr__(
+                    cls, "__post_init__", MaterialConstants.__post_init__
+                )
+
+    def to_units(
+        self: _MaterialConstants, units: pp.Units = pp.Units()
+    ) -> _MaterialConstants:
+        """Utility to quickly convert material constants to new units.
+
+        Note:
+            The default value of ``units`` allows a conversion to SI units.
+            When using units which are not defined in the standard
+            :class:`~porepy.models.units.Units` class, the default value of ``units``
+            might not be enough.
```

Comment:
The user must be aware which `Units` object was passed at instantiation.
If it was one which declared additional units to the ones defined in the base `Units` class, the conversion to SI (with the default value of `units`) will fail because it is the base class.

So if you have some `MyUnits` class, which was passed like `f = FluidConstants(units=MyUnits(...))`, then you also need `f.to_units(MyUnits())` to convert `f` back to SI (if you had non-SI when instantiating `f`)

Also, this default value here matches the default value of the `f.units` field.

I am open to not providing a default value here, to force the user to remember whether custom units classes were used or not.

Reply:
Perhaps removing the default is a good idea. Cause it's not going to hurt much (you can always pass pp.Units()). I also am a bit afraid of having a default which is an instance, which, I believe, will only be instantiated once when importing (right?). Also, do you think this note might just as well go into the parameter description, since it relates strictly to the parameter ``units``?

Reply:
Removed the default value.